client-data/tools/image/image.js

```diff
@@ -0,0 +1,276 @@
+
+/**
+ *                        WHITEBOPHIR
+ *********************************************************
+ * @licstart  The following is the entire license notice for the
+ *  JavaScript code in this page.
+ *
+ * Copyright (C) 2013  Ophir LOJKINE
+ *
+ *
+ * The JavaScript code in this page is free software: you can
+ * redistribute it and/or modify it under the terms of the GNU
+ * General Public License (GNU GPL) as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.  The code is distributed WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.
+ *
+ * As additional permission under GNU GPL version 3 section 7, you
+ * may distribute non-source (e.g., minimized or compacted) forms of
+ * that code without the copy of the GNU GPL normally required by
+ * section 4, provided you include this license notice and a URL
+ * through which recipients can access the Corresponding Source.
+ *
+ * @licend
+ */
+
+(function () { //Code isolation
+  const drawingArea = document.getElementById('board');
+  const canvas = document.getElementById('canvas');
+  const newImageDropPoint = {
+    x: 0,
+    y: 0,
+  };
+
+  const fileInput = document.createElement('input');
+  fileInput.type = 'file';
+  fileInput.accept = 'image/*';
+
+  function onFileInputChange(event) {
+    uploadImage(event.target.files[0], {
+      x: newImageDropPoint.x,
+      y: newImageDropPoint.y,
+    });
+  }
+
+  function promptForImage(x, y, event) {
+    // Get the position of the click on the canvas so when the user uploads
+    // an image, we can draw it at the same position.
+    newImageDropPoint.x = x;
+    newImageDropPoint.y = y;
+    fileInput.click();
+  }
+
+  function draw(data) {
+    Tools.drawingEvent = true;
+    switch (data.type) {
+      case "image":
+        createImageElement(data);
+        break;
+      case "update":
+        var image = svg.getElementById(data['id']);
+        if (!image) {
+          console.error("Image: No image provided!", data['id']);
+        }
+        updateImageElement(image, data);
+        break;
+      default:
+        console.error("Image: Draw instruction with unknown type. ", data);
+        break;
+    }
+  }
+
+  var svg = Tools.svg;
+  /**
+    * Creates a new image element on the canvas, or updates an existing image
+    * with new information.
+    * @param {Object} data - The data to use to create the image.
+    */
+  function createImageElement(data) {
+    var img = svg.getElementById(data.id) || Tools.createSVGElement("image");
+    img.setAttribute("id", data.id);
+    img.setAttribute("href", data.src);
```

Comment:
This is a big security vulnerability, isn't it ?

Reply:
Anyone can send an image event with the `src` attribute they want

Reply:
Is the concern that a user of Board A would be able to load images from Board B (which may be a "private" board)?

Reply:
The bigger concern is that they may have all users make requests to their own server

Reply:
Ah OK - so the concern is that an attacker could inject a URL that points to a server that they control, and that serves files with malicious content.  Is that right?

Reply:
Yes, they could serve whatever they want, but also log the ip addresses of everyone connected

Reply:
I believe this has been addressed with the latest commit 👍 Let me know if you disagree.